sdks/python/apache_beam/coders/coders.py

```diff
@@ -1383,22 +1375,74 @@ def from_runner_api_parameter(payload, components, context):
         write_state_threshold=int(payload))
 
 
-class RunnerAPICoderHolder(Coder):
+class ElementTypeHolder(typehints.TypeConstraint):
+  """A dummy element type for external coders that cannot be parsed in Python"""
+  def __init__(self, coder, context):
+    self.coder = coder
+    self.context = context
+
+
+class ExternalCoder(Coder):
   """A `Coder` that holds a runner API `Coder` proto.
 
   This is used for coders for which corresponding objects cannot be
   initialized in Python SDK. For example, coders for remote SDKs that may
   be available in Python SDK transform graph when expanding a cross-language
   transform.
   """
-  def __init__(self, proto):
-    self._proto = proto
 
-  def proto(self):
-    return self._proto
+  coder_count = 0
 
-  def to_runner_api(self, context):
-    return self._proto
+  def __init__(self, element_type_holder):
+    self.element_type_holder = element_type_holder
 
-  def to_type_hint(self):
-    return Any
+  def as_cloud_object(self, coders_context=None):
+    if not coders_context:
+      raise Exception(
+          'coders_context must be specified to correctly encode external coders'
+      )
+    coder_id = coders_context.get_by_proto(
+        self.element_type_holder.coder, deduplicate=True)
+
+    coder_proto = self.element_type_holder.coder
+
+    kind_str = 'kind:external' + str(ExternalCoder.coder_count)
+    ExternalCoder.coder_count = ExternalCoder.coder_count + 1
+    component_encodings = []
+    if coder_proto.spec.urn == 'beam:coder:kv:v1':
+      kind_str = 'kind:pair'
+      for component_coder_id in coder_proto.component_coder_ids:
+        component_encodings.append({
+            '@type': 'kind:external' + str(ExternalCoder.coder_count),
+            'pipeline_proto_coder_id': component_coder_id
+        })
+        ExternalCoder.coder_count = ExternalCoder.coder_count + 1
+
+    value = {
+        # This is a placeholder type. Dataflow will get the actual coder from
+        # pipeline proto using the pipeline_proto_coder_id property.
+        '@type': kind_str,
+        'pipeline_proto_coder_id': coder_id
+    }
+    if component_encodings:
+      value['is_pair_like'] = True
+      value['component_encodings'] = component_encodings
+
+    return value
+
+  @staticmethod
+  def from_type_hint(typehint, unused_registry):
+    if isinstance(typehint, ElementTypeHolder):
+      return ExternalCoder(typehint)
+    else:
+      raise ValueError((
+          'Expected an instance of ElementTypeHolder'
+          ', but got a %s' % typehint))
+
+  def to_runner_api_parameter(self, context):
+    if self.element_type_holder.coder.component_coder_ids:
+      raise NotImplementedError
```

Comment:
Why raise NotImplementedError here instead of returning this in place of the empty tuple below?